controllers/packetmachine_controller.go

```diff
@@ -164,6 +164,13 @@ func (r *PacketMachineReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		}
 	}()
 
+	// Add finalizer first if not set to avoid the race condition between init and delete.
```

Comment:
This refactor follows https://github.com/kubernetes-sigs/cluster-api/issues/3205